File: app/controllers/trainers_controller.rb
class TrainersController < ApplicationController
  before_action :authenticate_user!

  def buy
    base_cost = 100
    cost = (base_cost * (1.15 ** current_user.trainers)).to_i

    if current_user.money >= cost
      current_user.update!(
        money: current_user.money - cost,
        trainers: current_user.trainers + 1,
        auto_click_rate: current_user.trainers * 0.1
      )

      new_cost = (base_cost * (1.15 ** current_user.trainers)).to_i

      render json: {
        money: current_user.money,
        cost: cost,
        trainers: current_user.trainers,
        next_cost: new_cost
      }
    else
      render json: { error: "Not enough money" }, status: :unprocessable_entity
    end
  end
end

File: app/javascript/controllers/trainers_controller.js
// app/javascript/controllers/trainers_controller.js
import { Controller } from "@hotwired/stimulus"

export default class extends Controller {
  static targets = ["button"]

  async buyTrainer() {
    this.buttonTarget.disabled = true

    const response = await fetch("/trainers/buy", {
      method: "POST",
      headers: {
        "X-CSRF-Token": document.querySelector("[name='csrf-token']").content,
        "Content-Type": "application/json"
      },
      body: JSON.stringify({})
    }).then(() => {
      this.refreshFunctions()
    })

    if (response.ok) {
      const data = await response.json()
      this.buttonTarget.textContent = `Hire Trainer (Cost: ${data.next_cost})`
    } else {
      alert("Not enough money or an error occurred.")
    }

    this.buttonTarget.disabled = false
  }

  refreshFunctions() {
    fetch("/encounters/functions")
      .then(response => response.text())
      .then(html => {
        const statsElement = document.getElementById("functions")
        if (statsElement) {
          statsElement.innerHTML = html
        }
      })
  }
}
